builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1030,392 @@ func (b *backend) pathLoginRenew(
+			shortestTTL = maxTTL
+		}
+
+		resp.Auth.TTL = shortestTTL
+	}
+
+	return resp, nil
+}
+
+// These two methods (hasValuesFor*) return two bools
+// The first is a hasAll, that is, does the request have all the values
+// necessary for this auth method
+// The second is a hasAny, that is, does the request have any of the fields
+// exclusive to this auth method
+func hasValuesForEc2Auth(data *framework.FieldData) (bool, bool) {
+	_, hasPkcs7 := data.GetOk("pkcs7")
+	_, hasIdentity := data.GetOk("identity")
+	_, hasSignature := data.GetOk("signature")
+	return (hasPkcs7 || (hasIdentity && hasSignature)), (hasPkcs7 || hasIdentity || hasSignature)
+}
+
+func hasValuesForIamAuth(data *framework.FieldData) (bool, bool) {
+	_, hasRequestMethod := data.GetOk("request_method")
+	_, hasRequestUrl := data.GetOk("request_url")
+	_, hasRequestBody := data.GetOk("request_body")
+	_, hasRequestHeaders := data.GetOk("request_headers")
+	return (hasRequestMethod && hasRequestUrl && hasRequestBody && hasRequestHeaders),
+		(hasRequestMethod || hasRequestUrl || hasRequestBody || hasRequestHeaders)
+}
+
+func parseIamArn(iamArn string) (string, string, string, error) {
+	fullParts := strings.Split(iamArn, ":")
+	principalFullName := fullParts[5]
+	parts := strings.Split(principalFullName, "/")
+	principalName := parts[1]
+	transformedArn := iamArn
+	sessionName := ""
+	if parts[0] == "assumed-role" {
+		transformedArn = fmt.Sprintf("arn:aws:iam::%s:role/%s", fullParts[4], principalName)
+		sessionName = parts[2]
+	} else if parts[0] != "user" {
+		return "", "", "", fmt.Errorf("unrecognized principal type: '%s'", parts[0])
+	}
+	return transformedArn, principalName, sessionName, nil
+}
+
+func ensureVaultHeaderValue(headers http.Header, requestUrl *url.URL, requiredHeaderValue string) (bool, string) {
+	providedValue := ""
+	for k, v := range headers {
+		if strings.ToLower(iamServerIdHeader) == strings.ToLower(k) {
+			providedValue = strings.Join(v, ",")
+			break
+		}
+	}
+	if providedValue == "" {
+		return false, fmt.Sprintf("didn't find %s", iamServerIdHeader)
+	}
+
+	// NOT doing a constant time compare here since the value is NOT intended to be secret
+	if providedValue != requiredHeaderValue {
+		return false, fmt.Sprintf("expected %s but got %s", requiredHeaderValue, providedValue)
+	}
+
+	if authzHeaders, ok := headers["Authorization"]; ok {
+		// authzHeader looks like AWS4-HMAC-SHA256 Credential=AKI..., SignedHeaders=host;x-amz-date;x-vault-awsiam-id, Signature=...
+		// We need to extract out the SignedHeaders
+		re := regexp.MustCompile(".*SignedHeaders=([^,]+)")
+		authzHeader := strings.Join(authzHeaders, ",")
+		matches := re.FindSubmatch([]byte(authzHeader))
+		if len(matches) < 1 {
+			return false, "vault header wasn't signed"
+		}
+		if len(matches) > 2 {
+			return false, "found multiple SignedHeaders components"
+		}
+		signedHeaders := string(matches[1])
```

Comment:
I am sorry, lack of state here. How do we know that the desired value is at index 1?